lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp

```diff
@@ -1654,6 +1660,99 @@ bool SymbolFileDWARF::CompleteType(CompilerType &compiler_type) {
   return false;
 }
 
+DWARFDIE SymbolFileDWARF::FindDefinitionDIE(const DWARFDIE &die) {
+  auto def_die_it = GetDeclarationDIEToDefinitionDIE().find(die.GetDIE());
+  if (def_die_it != GetDeclarationDIEToDefinitionDIE().end())
+    return GetDIE(def_die_it->getSecond());
+
+  ParsedDWARFTypeAttributes attrs(die);
+  const dw_tag_t tag = die.Tag();
+  TypeSP type_sp;
+  Log *log = GetLog(DWARFLog::TypeCompletion | DWARFLog::Lookups);
+  if (log) {
+    GetObjectFile()->GetModule()->LogMessage(
+        log,
+        "SymbolFileDWARF({0:p}) - {1:x16}: {2} type \"{3}\" is a "
+        "forward declaration DIE, trying to find definition DIE",
+        static_cast<void *>(this), die.GetOffset(), DW_TAG_value_to_name(tag),
+        attrs.name.GetCString());
+  }
+  // We haven't parse definition die for this type, starting to search for it.
+  // After we found the definition die, the GetDeclarationDIEToDefinitionDIE()
+  // map will have the new mapping from this declaration die to definition die.
+  if (attrs.class_language == eLanguageTypeObjC ||
```

Comment:
I'm not sure how big of a problem this is (as `CompleteType` above already contains some clang-specific code), but moving some clang-specific code out of `DWARFASTParserClang` and into `SymbolFileDWARF` is less than ideal, as the latter is also used with non-clang/non-C languages. Would it be possible to keep this code `DWARFASTParserClang` somehow?

Reply:
Which part of this function is clang-specific code?

Reply:
Definitely anything that is gated on `language == eLanguageType*C*` checks, but a case could also be made that the whole declaration-to-definition mapping should be within the purview of the DWARFASTParser instance responsible for the specific language. Having the logic split between an ASTParser (a language-specific class) and SymbolFileDWARF (language-agnostic) forces all languages to work in the same way. I don't know how much of a problem would that be in practice, but it would definitely make the code nicer (single responsibility principle, and all) if the logic lived in a central place.

Reply:
I realize this is a bit fuzzy, and there are already some arguably clang-specific pieces of code in SymbolFileDWARF, but the main reason I am bringing this up is because this code (some of it at least) has previously been in DWARFASTParserClang, and now it isn't -- which seems like a regression to me.